src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp

```diff
@@ -0,0 +1,977 @@
+          size_t(height_group_range * height_group_size),
+          size_t(width_group_range * width_group_size),
+      },
+      sycl::range<3>{1, size_t(height_group_size), size_t(width_group_size)},
+      queue,
+      kfn);
+}
+
+void avg_pool3d_kernel(
+    const Tensor& input,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    bool ceil_mode,
+    bool count_include_pad,
+    c10::optional<int64_t> divisor_override,
+    Tensor& output) {
+  TensorArg output_arg{output, "output", 1};
+  TensorArg input_arg{input, "input", 2};
+
+  checkAllSameGPU(__func__, {output_arg, input_arg});
+
+  const int kT = safe_downcast<int, int64_t>(kernel_size[0]);
+  const int kH = kernel_size.size() == 1
+      ? kT
+      : safe_downcast<int, int64_t>(kernel_size[1]);
+  const int kW = kernel_size.size() == 1
+      ? kT
+      : safe_downcast<int, int64_t>(kernel_size[2]);
+
+  const int dT = stride.empty() ? kT : safe_downcast<int, int64_t>(stride[0]);
+  const int dH = stride.empty() ? kH
+      : stride.size() == 1      ? dT
+                                : safe_downcast<int, int64_t>(stride[1]);
+  const int dW = stride.empty() ? kW
+      : stride.size() == 1      ? dT
+                                : safe_downcast<int, int64_t>(stride[2]);
+
+  const int padT = safe_downcast<int, int64_t>(padding[0]);
+  const int padH =
+      padding.size() == 1 ? padT : safe_downcast<int, int64_t>(padding[1]);
+  const int padW =
+      padding.size() == 1 ? padT : safe_downcast<int, int64_t>(padding[2]);
+
+  // if divisor==0 then we will ignore it
+  int64_t divisor = 0;
+  if (divisor_override.has_value()) {
+    divisor = divisor_override.value();
+  }
+
+  const int64_t nbatch = input.ndimension() == 5 ? input.size(-5) : 1;
+  const int64_t nslices = input.size(-4);
+  const int64_t itime = input.size(-3);
+  const int64_t iheight = input.size(-2);
+  const int64_t iwidth = input.size(-1);
+
+  const int64_t otime =
+      pooling_output_shape<int64_t>(itime, kT, padT, dT, 1, ceil_mode);
+  const int64_t oheight =
+      pooling_output_shape<int64_t>(iheight, kH, padH, dH, 1, ceil_mode);
+  const int64_t owidth =
+      pooling_output_shape<int64_t>(iwidth, kW, padW, dW, 1, ceil_mode);
+
+  Tensor work_input = input.contiguous();
+  Tensor work_output = output;
+
+  if (input.ndimension() == 5) {
+    // Collapse batch and feature dimensions.
+    work_input = work_input.reshape({nbatch * nslices, itime, iheight, iwidth});
+    work_output =
+        work_output.reshape({nbatch * nslices, otime, oheight, owidth});
+  }
+  AT_DISPATCH_FLOATING_TYPES_AND2(
+      kHalf, kBFloat16, input.scalar_type(), "avg_pool3d_out_xpu", [&] {
+        using accscalar_t = acc_type_device<scalar_t, kXPU>;
+        int64_t totalZ = otime * nslices * nbatch;
+        int64_t offsetZ = 0;
+        while (totalZ > 0) {
+          if (canUse32BitIndexMath(input)) {
+            avg_pool3d_out_frame<scalar_t, accscalar_t, int32_t>(
```

Comment:
CUDA avg_pool3d_out_frame means a CUDA kernel, not a launcher, here we can name is avg_pool3d_out_template or launch_avg_pool3d_kernel